File: backend/api/dome.py
from fastapi import APIRouter, HTTPException
from Controllers.Dome_Controller import DomeController

router = APIRouter(
    prefix="/api/dome",
    tags=["Dome"]
)

controller: DomeController | None = None


def set_controller(dome_controller: DomeController):
    global controller
    controller = dome_controller

def get_controller() -> DomeController:
    if controller is None:
        raise HTTPException(
            status_code=503,
            detail='Dome controller not initialised'
        )
    return controller

@router.post('/connect')
def connect():
    dome_controller = get_controller()

    result = dome_controller.connect()

    if not result:
        raise HTTPException(
            status_code=503,
            detail='Unable to connect to dome'
        )

    return {
        'success': True
    }

@router.post('/disconnect')
def disconnect():
    dome_controller = get_controller()

    dome_controller.disconnect()

    return {
        'success': True
    }

@router.get("/status")
def status():
    dome_controller = get_controller()
    return dome_controller.get_status()

@router.post('/open')
def open_dome():
    dome_controller = get_controller()

    result = dome_controller.open_dome()

    return {
        'success': result is not False
    }

@router.post('/close')
def close_dome():
    dome_controller = get_controller()

    result = dome_controller.close_dome()

    return {
        'success': result is not False
    }

@router.post('/open_one')
def open_one():
    dome_controller = get_controller()

    result = dome_controller.open_left()

    return {
        'success': result is not False
    }


@router.post('/close_one')
def close_one():
    dome_controller = get_controller()

    result = dome_controller.close_left()

    return {
        'success': result is not False
    }

@router.post('/open_two')
def open_two():
    dome_controller = get_controller()

    result = dome_controller.open_right()

    return {
        'success': result is not False
    }


@router.post('/close_two')
def close_two():
    dome_controller = get_controller()

    result = dome_controller.close_right()

    return {
        'success': result is not False
    }

File: frontend/ogs-gui/src/api/dome.ts
export interface DomeStatusData {
    connected: boolean;
    open: boolean;
    closed: boolean;
    opening: boolean;
    closing: boolean;
    moving: boolean;
    fault: boolean;
}


interface DomeStatusResponse {
    success: boolean;
    data: DomeStatusData;
}


async function getErrorMessage(
    response: Response,
    fallback: string
): Promise<string> {

    try {

        const data =
            await response.json();

        return (
            typeof data?.detail === "string"
                ? data.detail
                : fallback
        );

    } catch {

        return fallback;

    }
}


async function domePost(
    endpoint: string,
    fallback: string
): Promise<void> {

    const response =
        await fetch(
            `/api/dome/${endpoint}`,
            {
                method: "POST",
            }
        );


    if (!response.ok) {

        const message =
            await getErrorMessage(
                response,
                fallback
            );

        throw new Error(message);

    }

}


export async function getDomeStatus():
Promise<DomeStatusData> {

    const response =
        await fetch(
            "/api/dome/status"
        );


    if (!response.ok) {

        const message =
            await getErrorMessage(
                response,
                "Failed to get dome status"
            );

        throw new Error(message);

    }


    const result:
        DomeStatusResponse =
            await response.json();


    return result.data;

}


export function connectDome():
Promise<void> {

    return domePost(
        "connect",
        "Failed to connect dome"
    );

}


export function disconnectDome():
Promise<void> {

    return domePost(
        "disconnect",
        "Failed to disconnect dome"
    );

}


export function openDome():
Promise<void> {

    return domePost(
        "open",
        "Failed to open dome"
    );

}


export function closeDome():
Promise<void> {

    return domePost(
        "close",
        "Failed to close dome"
    );

}


export function openDomeOne():
Promise<void> {

    return domePost(
        "open_one",
        "Failed to open left shutter"
    );

}


export function closeDomeOne():
Promise<void> {

    return domePost(
        "close_one",
        "Failed to close left shutter"
    );

}


export function openDomeTwo():
Promise<void> {

    return domePost(
        "open_two",
        "Failed to open right shutter"
    );

}


export function closeDomeTwo():
Promise<void> {

    return domePost(
        "close_two",
        "Failed to close right shutter"
    );

}

